### PR TITLE
Consensus ledger switch improvements

### PR DESCRIPTION
**Pull request by JoelKatz**
* Expire validations faster based on when we first saw them.

* Never jump to a ledger prior to the latest fully-valid ledger

* Drop validations with signing times too far in the future immediately

### CODE REVIEW

**Comment by miguelportilla** [comment]
Missing an include but otherwise looks good. :+1:

File: src/ripple/app/ledger/LedgerTiming.h [hunk 1]
```diff
@@ -74,6 +74,11 @@ const int LEDGER_IDLE_INTERVAL = 15;
 // it takes to adjust the close time accuracy window
 const int LEDGER_VAL_INTERVAL = 300;
 
+// The number of seconds a validation remains current after the time we first
+// saw it. This provides faster recovery in very rare cases where the number
+// of validations produced by the network is lower than normal
+const int LEDGER_VAL_LOCAL = 180;
+
```
Comment (seelabs) [inline src/ripple/app/ledger/LedgerTiming.h]: `LEDGER_VAL_LOCAL` is too generic a name. `LEDGER_VAL_INTERVAL` deals with sign time, and `LEDGER_VAL_LOCAL` deal with seen time. How about renaming these `LEDGER_INTERVAL_{SIGN,SEEN}`?

File: src/ripple/app/misc/Validations.cpp [hunk 2]
```diff
@@ -346,9 +347,12 @@ class ValidationsImp : public Validations
     }
 
     LedgerToValidationCounter getCurrentValidations (
-        uint256 currentLedger, uint256 priorLedger) override
+        uint256 currentLedger,
+        uint256 priorLedger,
+        LedgerIndex cutoffBefore) override
     {
         auto const cutoff = app_.timeKeeper().now().time_since_epoch().count() - LEDGER_VAL_INTERVAL;
+        auto const lCutoff = cutoff + LEDGER_VAL_INTERVAL - LEDGER_VAL_LOCAL;
```
Comment (seelabs) [inline src/ripple/app/misc/Validations.cpp]: I think it's clearer if both `cutoff` and `lCutoff` were calculated from  a local `now` variable. I.e `cutoff = now -LEDGER_VAL_INTERVAL;`, `lCutoff = now - LEDGER_VAL_LOCAL;`

File: src/ripple/app/misc/Validations.cpp [hunk 3]
```diff
@@ -75,7 +75,7 @@ class ValidationsImp : public Validations
 public:
     ValidationsImp (Application& app)
         : app_ (app)
-        , mValidations ("Validations", 128, 600, stopwatch(),
+        , mValidations ("Validations", 4096, 600, stopwatch(),
```
Comment (seelabs) [inline src/ripple/app/misc/Validations.cpp]: `ApplicationImp::setup` calls `mValidations->tune` (line 1008), and sets the size from the config file. It looks like the place to really change this value is in `Config::getSize`, not here.

**Comment by scottschurr** [comment]
This looks good as far as I can see.  I left an entirely optional suggestion.  :+1:

File: src/ripple/app/misc/Validations.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/app/misc/Validations.cpp]: Agreed. Subtracting out `LEDGER_VAL_INTERVAL` only to add it back is confusing, wasteful, and is going to be fragile in the future.

**Comment by ximinez** [comment]
Looks good overall. :+1:

**Comment by ximinez** [comment]
Looks good overall, though I concur with @seelabs's suggestions. Otherwise :+1:

**Comment by nbougalis** [comment]
:+1:

**Comment by seelabs** [comment]
:+1:

**Comment by ximinez** [comment]
:+1:

**Pull request closed without merging**